File: 207-course-schedule/207-course-schedule.java
class Solution {
    
    public boolean dfs(int[]vis,ArrayList<Integer>[]graph,int src){
        vis[src]=1;
        
        for(int nbr:graph[src]){
            if(vis[nbr]==0){
                if(dfs(vis,graph,nbr))return true;
            }else if(vis[nbr]==1){
                return true;
            }
        }
        
        vis[src]=2;
        return false;
    }
    
    public boolean canFinish(int N, int[][] edges) {
        int[]vis=new int[N];
        ArrayList<Integer>[]graph=new ArrayList[N];
        
        for(int i=0;i<N;i++){
            graph[i]=new ArrayList<>();
        }
        
        for(int []edge:edges){
            int u=edge[1];
            int v=edge[0];
            
            graph[u].add(v);
        }
        
        for(int i=0;i<N;i++){
            if(vis[i]==0){
                boolean cycle=dfs(vis,graph,i);
                if(cycle)return false;
            }
        }
        
        return true;
    }
}

File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& edges) {
        vector<vector<int>>graph(n);
        vector<int>indegree(n,0);
        
        for(vector<int>&e:edges){
            graph[e[0]].push_back(e[1]);
            indegree[e[1]]++;
        }
        
        queue<int>q;
        for(int i=0;i<n;i++)if(indegree[i]==0)q.push(i);
        
        int vertexCount=0;
        
        while(q.size()!=0){
            int size=q.size();
            while(size-->0){
                int vtx=q.front();q.pop();
                vertexCount++;
                
                for(int v:graph[vtx]){
                    if(--indegree[v]==0){
                        q.push(v);
                    }
                }
            }
        }
        
        return vertexCount==n;
    }
};